File: CommonFiles/Common_Files_VB/Wrapper_Populated.vb
Imports Microsoft.VisualBasic
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.Immutable
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports System.Text

Namespace MetalUp.Express

    Module Program
        Public Sub Main()
            Console.OutputEncoding = Encoding.UTF8
            Console.WriteLine(Display(""))
        End Sub

        ' StudentCode - this comment is needed error line number offset
        Public Function IsPrime(ByVal n As Integer) As Boolean
            Return Not Enumerable.Range(2, n \ 2 - 1).Any(Function(f) IsFactor(f, n))
        End Function

#Region "<Helpers>"
        Public Function Display(ByVal obj As Object) As String
            If obj Is Nothing Then
                Return ""
            End If
            If TypeOf obj Is String Then
                Return $"{obj}"
            End If
            If TypeOf obj Is Boolean Then
                Return If(DirectCast(obj, Boolean), "true", "false")
            End If
            Dim type = obj.GetType()
            If TypeOf obj Is Tuple Then
                Dim toDisplay = DirectCast(obj, IEnumerable).Cast(Of Object)().Select(Function(o) Display(o)).ToList()
                Return $"({String.Join(","c, toDisplay)})"
            End If
            If type.IsGenericType Then
                If type.GetGenericTypeDefinition() Is GetType(List(Of )) Then
                    Dim toDisplay = DirectCast(obj, IEnumerable).Cast(Of Object)().Select(Function(o) Display(o)).ToList()
                    Return $"{{{String.Join(","c, toDisplay)}}}"
                End If
            End If
            If TypeOf obj Is IEnumerable Then
                Return "Result is an IEnumerable. Convert to List to display contents"
            End If
            Return obj.ToString()
        End Function

        Public Function ArgString(ParamArray arguments As Object()) As String
            Return arguments.Aggregate("", Function(s, a) s & Display(a) & ", ").TrimEnd(" "c, ","c)
        End Function

        Public Function FailMessage(functionName As String, ByVal expected As Object, ByVal actual As Object, ParamArray args As Object()) As String
            Return $"xxxTest failed calling {functionName}({ArgString(args)}) Expected: {Display(expected)} Actual: {Display(actual)}xxx"
        End Function

        Public Function QAFailMessage(ByVal wrongAnswers As String) As String
            Return $"xxxWrong or missing answer(s) to question number(s): {wrongAnswers}.xxx"
        End Function

        Public Function WrongAnswers(ByVal student As String, ByVal answers As String) As String
            Dim studentArr = AsArray(student)
            Dim answersArr = AsArray(answers)
            Dim result = ""
            For n As Integer = 0 To answersArr.Length - 1
                If answersArr(n) IsNot Nothing AndAlso studentArr(n) <> answersArr(n) Then
                    result &= $"{n} "
                End If
            Next n
            Return result
        End Function

        Public Function AsArray(ByVal answers As String) As String()
            Dim arr = New String(19) {}
            For Each a As String In answers.Split(ControlChars.Lf).Where(Function(x) x.Length > 0)
                Dim split = a.Split(")"c)
                Dim n = AsInt(split(0).Trim())
                arr(n) = split(1).Trim().ToUpper()
            Next a
            Return arr
        End Function

        Private Function AsInt(ByVal input As String) As Integer
            Return Convert.ToInt32(Encoding.Default.GetString(Encoding.ASCII.GetBytes(input).Where(Function(b) b > 47 AndAlso b < 58).ToArray()))
        End Function

        <Extension>
        Public Function SetItem(Of T)(ByVal list As List(Of T), ByVal i As Integer, ByVal value As T) As List(Of T)
            Return list.Take(i).Append(value).Concat(list.Skip(i + 1)).ToList()
        End Function

        <Extension>
        Public Function InsertItem(Of T)(ByVal list As List(Of T), ByVal i As Integer, ByVal value As T) As List(Of T)
            Return list.Take(i).Append(value).Concat(list.Skip(i)).ToList()
        End Function

        <Extension>
        Public Function RemoveItem(Of T)(ByVal list As List(Of T), ByVal i As Integer) As List(Of T)
            Return list.Take(i).Concat(list.Skip(i)).ToList()
        End Function

#End Region

#Region "<HiddenCode>"
        Public Function IsFactor(ByVal f As Integer, ByVal n As Integer) As Boolean
            Return n Mod f = 0
        End Function
#End Region

    End Module

    '<Tests>

End Namespace

File: CommonFiles/Common_Files_CS/Wrapper_Populated.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text;
using static MetalUp.Express.Extensions;

namespace MetalUp.Express
{
    public class Wrapper_CS
    {
        static void Main()
        {
            Console.OutputEncoding = Encoding.UTF8;
            Console.WriteLine(Display(3));
        }

        // StudentCode - this comment is needed error line number offset
        //<StudentCode>

        //<HiddenCode>

        #region <Helpers>
        public static string Display(object obj)
        {
            if (obj is null) return "";
            if (obj is string) return $@"{obj}";
            if (obj is Boolean) return (Boolean)obj ? "true" : "false";
            var type = obj.GetType();
            if (type == typeof(Tuple))
            {
                var toDisplay = ((IEnumerable)obj).Cast<object>().Select(o => Display(o)).ToList();
                return $@"({string.Join(',', toDisplay)})";
            }
            if (type.IsGenericType)
            {
                if (type.GetGenericTypeDefinition() == typeof(List<>))
                {
                    var toDisplay = ((IEnumerable)obj).Cast<object>().Select(o => Display(o)).ToList();
                    return $@"{{{string.Join(',', toDisplay)}}}";
                }
            }
            if (obj is IEnumerable)
            {
                return "Result is an IEnumerable. Convert to List to display contents";
            }
            return obj.ToString();
        }

        public static string ArgString(params object[] arguments) => arguments.Aggregate("", (s, a) => s + Display(a) + ", ").TrimEnd(' ', ',');

        public static string FailMessage(string functionName, object expected, object actual, params object[] args) =>
             $"xxxTest failed calling {functionName}({ArgString(args)}) Expected: {Display(expected)} Actual: {Display(actual)}xxx";

        public static string QAFailMessage(string wrongAnswers) =>
            $"xxxWrong or missing answer(s) to question number(s): {wrongAnswers}.xxx";

        public static string WrongAnswers(string student, string answers)
        {
            var studentArr = AsArray(student);
            var answersArr = AsArray(answers);
            var result = "";
            for (int n = 0; n < answersArr.Length; n++)
            {
                if (answersArr[n] != null && studentArr[n] != answersArr[n]) result += $"{n} ";
            }
            return result;
        }

        public static string[] AsArray(string answers)
        {
            var arr = new string[20];
            foreach (var a in answers.Split('\n').Where(a => a.Length > 0))
            {
                var split = a.Split(')');
                var n = AsInt(split[0].Trim());
                arr[n] = split[1].Trim().ToUpper();
            }
            return arr;
        }

        private static int AsInt(string input) =>
             Convert.ToInt32(Encoding.Default.GetString(Encoding.ASCII.GetBytes(input).Where(b => b > 47 && b < 58).ToArray()));


        #endregion
        //<Tests>
    }

    public static class Extensions
    {
            public static List<T> SetItem<T>(this List<T> list, int i, T value) =>
                list.Take(i).Append(value).Concat(list.Skip(i + 1)).ToList();

            public static List<T> InsertItem<T>(this List<T> list, int i, T value) =>
                list.Take(i).Append(value).Concat(list.Skip(i)).ToList();
    }
}